Returned exchange item should cancel charge of unreturned original item

gvaughn opened the pull request:
I'm going with a new 'unexchanged' status of the original ReturnItem.
I've added :unexchanged as a COMPLETED_RECEPTION_STATUSES so it won't get picked up with exchanges.rake. No tests fail, and it seems reasonable to be treated like :received, but still needs more brainpower devoted to it.

gvaughn commented:
I updated syntax, so i might be helpful to look at individual commits

gvaughn commented on core/spec/models/spree/order/checkout_spec.rb:
```diff
@@ -628,6 +628,7 @@ class SubclassedOrder < Spree::Order
 
       order.payments.create!({ :amount => order.outstanding_balance, :payment_method => payment_method, :source => creditcard })
       order.complete!
+      #TODO why is this test failing?
```
This is my current conundrum that I'll be tackling after lunch

athal7 commented on core/app/models/spree/return_item.rb:
```diff
@@ -188,6 +190,14 @@ def process_inventory_unit!
     end
 
+    def check_unexchange
+      if unexchanged_sibling = self.class.find_by(exchange_inventory_unit: inventory_unit)
```
i think we probably want to only look for awaiting here, otherwise we might be unexchanging successful exchanges

athal7 commented on core/app/models/spree/return_item.rb:
```diff
@@ -63,11 +63,13 @@ class ReturnItem < ActiveRecord::Base
       event(:cancel) { transition to: :cancelled, from: :awaiting }
 
       event(:receive) { transition to: :received, from: INTERMEDIATE_RECEPTION_STATUSES + [:awaiting] }
+      event(:unexchange) { transition to: :unexchanged, from: INTERMEDIATE_RECEPTION_STATUSES + [:awaiting] }
```
i think only transitioning from awaiting here would be appropriate, same reasoning as below

gvaughn commented:
@athal7 I took care of those 2 awaiting state things. And now I'm back looking at that checkout_spec failure (mentioned in the description above)

gmacdougall commented on core/app/models/spree/return_item.rb:
```diff
@@ -188,6 +190,14 @@ def process_inventory_unit!
+      if unexchanged_sibling = self.class.awaiting_return.find_by(exchange_inventory_unit: inventory_unit)
+        # This happens when we ship an exchange to a customer, but the customer keeps the original and returns the exchange
+        unexchanged_sibling.unexchange!
+        update_attributes(pre_tax_amount: 0)
```
I am confused by this. Could you explain why this is the correct thing to do here?

gvaughn replied:
I'm glad you asked. Since this is my first task (and Andrew did not give me a _simple_ one) I'll explain how I understand it, which is wide open to correction. Here's a "for example":

* A customer buys pants with a 32 inch inseam -- paid and delivered
* They ask to exchange for 34 length, which we ship to them and ask they return the 32s within 3 weeks
* We do some bookkeeping here, which is still a bit fuzzy to me. This return item has a pre_tax_amount that we don't charge right away, but a cron job will pick it up and charge if the customer does not return the original within 3 weeks.
* The customer tries on the new pants and decides he prefer's the original 32 length, so he returns the 34s
* In 3 weeks the current cron job will note that the original 32 pants were not returned and charge the customer for the 34s assuming he kept both of them.
* This code notes when the 34s arrive, that they were an exchange for the 32s, the customer still only has 1 pair of pants, so we do not charge for a 2nd pair, and we put the 32s in a new 'unexchanged' state.

athal7 replied:
@gvaughn that explanation is great as far as the original item is concerned, the other half of this is that we would immediately reimburse the new (returned) item for the full amount. setting the pre_tax_amount to 0 will make the reimbursement safe to perform (so it doesn't appear to be pending), but also not give the customer back any money (which is correct, since they have just sent back their new item and we are re-enabling the old item to be returned by 'unexchanging' it)

athal7 replied:
with that said, i'm not sure we are best off writing the pre_tax_amount here, i think a superior approach could be to hook into the 'refund_amount_calculator' with this logic and calling 'set_default_pre_tax_amount' again so that this can be configured

athal7 commented:
lgtm!

gvaughn commented:
I'm fighting strange test behavior. I've tried running the tests with rake and with rspec directly. I've done full suites, individual tests, and a focus on the new test plus the one that breaks. Here's the results:

full rake == fail (message 'authorize!' was received by one Spree::Payment, but has already been received by another Spree::Payment -- in the checkout_spec:621)
full rspec == pass
singular rake (with SPEC env var) == fail (same failure message as above)
singular rspec == pass
focus rake (with both) == pass
focus rspec (with both) == pass

If I focus in on the line "update_attributes(pre_tax_amount: 0)" and replace with "pre_tax_amount = 0" then in the singular rake case, the checkout_spec passes, but the new return_item_spec fails (in all ways of launching)

@athal7 Do you have time to dig into this with me this afternoon?

athal7 commented:
Ah! We are in a state machine transition that will commit after it finishes processing, so assignment without the transaction would be a better fit. Happy to look with you

gvaughn commented:
I have moved the logic of price setting into the Calculator, which I prefer, now that I've pieced together how it works. I've rebased and done a bit of rspec focus config. It's ready for another review.

One thing still bugs me though. If I run the full suite with `rake` everything passes. If I run it as `rspec` then that same checkout_spec:628 fails. But I suppose that oddness is best left for another day.

athal7 commented on core/app/models/spree/return_item.rb:
```diff
@@ -162,6 +164,11 @@ def potential_reception_transitions
       status_paths.map{ |s| s.to_s.humanize }.zip(event_paths)
     end
 
+    def sibling_intended_for_exchange
```
nice!

athal7 commented on core/app/models/spree/return_item.rb:
```diff
@@ -188,6 +195,10 @@ def process_inventory_unit!
       customer_return.process_return! if customer_return
     end
 
+    def check_unexchange
```
i'm surprised that we don't need to tell it to recalculate the pre_tax amount here.

gvaughn replied:
I saw that it was called several times by a `puts`. Upon closer inspection it's also called from ReturnItem.from_inventory_unit. But that seems to be only when created too.

OK, now you've given me a serious head scratch moment

gvaughn replied:
Upon further research, I _think_ this is OK, but I'm less than fully confident. I am guessing the overall process goes:

* we physically receive the exchange item
* somehow that gets looked up as an inventory item
* a ReturnItem.from_inventory_item call is made to find or create the ReturnItem

If that is the case, then my test is less than accurate because I'm using the fixture to create it.

Thoughts?

athal7 replied:
i think that is one valid use case, though i think it is also possible that this is being received multiple days after being created, so i think recalculation would be a good choice

gvaughn replied:
I'd like to talk more about this. Seems to me that the pricing of the returned exchange item can be set to 0 upon creation. A delay until the original item is marked unexchanged shouldn't hurt anything based upon my understanding.

gvaughn commented:
Yet another attempt to get this working.

And I see that the checkout_spec:628 problem is back. I'm trying to figure that one out now.

gvaughn commented:
Wait. CircleCI passed !?!

athal7 commented:
:+1: from me. Good work. Just need a review from @bonobos/free-running-technologies

gmacdougall commented on core/app/models/spree/calculator/returns/default_refund_amount.rb:
```diff
@@ -9,7 +9,7 @@ def self.description
 
       def compute(return_item)
-        return 0.0.to_d if return_item.exchange_requested?
+        return 0.0.to_d if return_item.exchange_requested? || return_item.sibling_unexchanged?
```
I might suggest:

```Ruby
return 0.to_d unless return_item.eligible_for_refund?
```

And then within `Spree::ReturnItem`:

```
def eligible_for_refund?
  !exchange_requested? && !sibling_unexchanged?
end
```

This gives a couple simple improvements:

1. It's easier to write a test here, as move some of the responsibility to the `return_item`
2. In the future, if there are additional reasons why the `return_item` cannot be refunded, we can deal with that in a single place.

gvaughn replied:
Great! I love that idea. Plus, by doing that #sibling_unexchanged? can become private and avoid the concern below

gmacdougall replied:
Yup. You'll still want to bring `eligible_for_refund?` under test in that case though.

gmacdougall commented on core/app/models/spree/return_item.rb:
```diff
@@ -162,6 +164,10 @@ def potential_reception_transitions
       status_paths.map{ |s| s.to_s.humanize }.zip(event_paths)
     end
 
+    def sibling_unexchanged?
```
As this is an addition to the public API, this should have documentation and test coverage.

gvaughn commented:
@gmacdougall I implemented your suggestion under a slightly different name so things could remain positive and not need the `!`s.

All tests passed for me locally (which was a bit of a surprise, but a good surprise :-)

athal7 commented on core/app/models/spree/return_item.rb:
```diff
@@ -162,6 +164,12 @@ def potential_reception_transitions
       status_paths.map{ |s| s.to_s.humanize }.zip(event_paths)
     end
 
+    def part_of_exchange?
```
i find this name just a bit misleading. what do you think of `return_of_exchanged_item?` ?

gvaughn replied:
To my mind, that name focuses on the 2nd part of the conditional, but you have more experience in the codebase and domain terminology, so I'm happy to go with whatever you think best.

athal7 replied:
yea, i'm just concerned there are some cases that are part of an exchange that don't fall into this? but i guess they won't be actively part of an exchange if they don't hit this conditional. so I'm :+1: on this change

gvaughn replied:
Should we call it `exchange_in_progress?`

athal7 replied:
i'm happy with either that or the current. :shipit:

gmacdougall commented on core/spec/models/spree/return_item_spec.rb:
```diff
@@ -459,24 +477,35 @@
+    context "it's the exchange itself being returned" do
+      before { expect(subject).to receive(:sibling_intended_for_exchange).with('unexchanged').and_return(true) }
+      it { expect(subject.part_of_exchange?).to eq true }
+    end
```
What would cause part_of_exchange? to be false?

gvaughn replied:
Test case added

gmacdougall commented:
:+1:
